### ISSUE TITLE
Performance of train from scratch

### ISSUE THREAD

**Issue by deJQK** [ali-vilab/FACM#4]
Thanks again for sharing the insightful work. I noticed in the paper, most results are those with 2-stage training, i.e. FM pretraining followed by FACM distillation. Do you have the performance of training FACM from scratch in 1 stage, i.e., without FM pretraining? Thanks!

**Comment by deJQK** [comment]
Got it. Thanks. BTW, why did you closed the other one? They are different questions. Thanks.

**Comment by deJQK** [comment]
Hi, is it possible to share the model and training codebase for CIFAR10? How did you train with simple-EDM? It's okay if cleaning the code for that part is time consuming and not easy to share, but could you please share some more details on this? Thanks!

**Comment by Peterande** [comment]
Hi @deJQK. You can refer to sCM and IMM. In Section C.5, 'Simplified Parameterization', the IMM paper details how the Simple-EDM parameterization is equivalent to v-prediction when using the cosine noise schedule.

**Comment by deJQK** [comment]
Thanks @Peterande for the explanation. I meant if you have code for EDM on cifar with FACM. For example, did you directly use their model and preconditioning? In EDM, after reparameterization (like the steps you mentioned in IMM), the output is equivalent to v-pred for the network's output. Are you using vp with unconditional cifar? Besides those in Table 5 in your paper, are there any other settings that need to take care about to reproduce your results? Thanks again.

**Comment by deJQK** [comment]
I am trying to train FACM with edm model on cifar10, from scratch without distillation (because I would like to try something that cannot use the original edm model directly), which might take around 8 days (2.7 steps/sec, and totally 2M steps, which is 2e6/2.7/86400 ~ 8.5days). The FID for 120k steps is around 10.73 for 4 steps generation and 18.27 for 1 step generation. I understand the results reported in the paper might probably be that from distillation. Do you expect this results reasonable based on the previous experiments on ImageNet? Thanks!

**Comment by Peterande** [comment]
To the best of my knowledge, your FID @ 120k on Cifar10 seems higher than expected, and the convergence is a bit slow. How many epochs is approximately equivalent to your setting of 120k? The default design of FACM is more friendly to distillation because the idea is more about anchoring a fixed instantaneous velocity field, which we have proven on larger models (currently, the 14B scale works well). If you have to train from scratch, you can reduce the weight of the cm loss to accelerate the early-stage convergence. You can understand it this way: the cm task is built on the basis of the well-learned fm task, and the two different tasks in the early stage may not be conducive to the model's rapid initialization.

**Comment by deJQK** [comment]
Thanks very much @Peterande for your kind explanation and discussion, it is very precious for me and saves a lot of my time.

120k iters correspond to epoch 307, and I am using a batch size of 128 with 4 GPUs. The total number of epochs is 5120, which corresponds to 2M iters. Now for 380k iters (epoch 974), the FIDs are 11.26 for 1 step, 5.92 for 2 steps, and 5.82 for 4 steps. I have the same feeling that the convergence of FID might be slow. The plot of FID is like the following:

<img width="2400" height="1800" alt="Image" src="https://github.com/user-attachments/assets/f38b963c-274e-4c33-a302-f5f01e4fa133" />


The losses and grad norm look like the following:
<img width="2400" height="1800" alt="Image" src="https://github.com/user-attachments/assets/a503c1e7-5298-495e-b9b3-453190773ece" />

<img width="2400" height="1800" alt="Image" src="https://github.com/user-attachments/assets/ce2185d8-2cc4-450e-95cf-3e7c3e341a6b" />

<img width="2400" height="1800" alt="Image" src="https://github.com/user-attachments/assets/3bd1b967-b584-4960-878b-cf6300b4cdd9" />

It can be seen that loss_cm oscillate a lot (I have tried previously on some 2D toy data before and it seems loss_cm oscillation is a common issue).

Maybe I'd better to adopt a two-stage strategy to train a velocity model first, but that might be more costly and need more hyperparameter tuning (i.e., first tune the training of the velocity model). Your suggestion of adjusting CM loss weight is also reasonable, and it seems like curriculum training, which might also need some more tuning. Thanks for your insights of the CM loss and I might need some time to further understand this part.

Thanks very much again for your kind discussion and suggestion.

**Comment by deJQK** [comment]
Should it be 'emas' instead of 'ema' [here](https://github.com/ali-vilab/FACM/blob/main/train.py#L103)?

**Comment by deJQK** [comment]
It seems that some extra steps/scripts are required to do posthoc ema to get a checkpoint with key 'ema'? And inside the 'ema' the keys should be the parameter names of the model?

**Comment by Peterande** [comment]
> Should it be 'emas' instead of 'ema' [here](https://github.com/ali-vilab/FACM/blob/main/train.py#L103)?

The teacher here is trained using a regular EMA, with no involvement of posthoc EMA. Additionally, in practice, we only use the first model maintained by posthoc EMA and do not synthesize new models afterward

**Comment by deJQK** [comment]
> > Should it be 'emas' instead of 'ema' [here](https://github.com/ali-vilab/FACM/blob/main/train.py#L103)?
> 
> The teacher here is trained using a regular EMA, with no involvement of posthoc EMA. Additionally, in practice, we only use the first model maintained by posthoc EMA and do not synthesize new models afterward

Thanks a lot for your kind explanation. Is this correct that the posthoc ema is only used for the two-stage case of FACM, and applied on the distilled student?
